src/main/java/io/debezium/connector/jdbc/JdbcChangeEventSink.java

```diff
@@ -0,0 +1,406 @@
+            sessionFactory.close();
+        }
+        else {
+            LOGGER.info("Session factory already closed");
+        }
+    }
+
+    private TableDescriptor checkAndApplyTableChangesIfNeeded(SinkRecord record, SinkRecordDescriptor descriptor) throws SQLException {
+        final String tableName = config.getTableNamingStrategy().resolveTableName(config, record);
+        if (!hasTable(tableName)) {
+            // Table does not exist, lets attempt to create it.
+            try {
+                return createTable(tableName, descriptor);
+            }
+            catch (SQLException ce) {
+                // It's possible the table may have been created in the interim, so try to alter.
+                LOGGER.warn("Table creation failed for '{}', attempting to alter the table", tableName, ce);
+                try {
+                    return alterTableIfNeeded(tableName, descriptor);
+                }
+                catch (SQLException ae) {
+                    // The alter failed, hard stop.
+                    LOGGER.error("Failed to alter the table '{}'.", tableName, ae);
+                    throw ae;
+                }
+            }
+        }
+        else {
+            // Table exists, lets attempt to alter it if necessary.
+            try {
+                return alterTableIfNeeded(tableName, descriptor);
+            }
+            catch (SQLException ae) {
+                LOGGER.error("Failed to alter the table '{}'.", tableName, ae);
+                throw ae;
+            }
+        }
+    }
+
+    private boolean hasTable(String tableName) {
+        return session.doReturningWork((connection) -> dialect.tableExists(connection, tableName));
+    }
+
+    private TableDescriptor readTable(String tableName) {
+        return session.doReturningWork((connection) -> dialect.readTable(connection, tableName));
+    }
+
+    private TableDescriptor createTable(String tableName, SinkRecordDescriptor record) throws SQLException {
+        LOGGER.debug("Attempting to create table '{}'.", tableName);
+
+        if (NONE.equals(config.getSchemaEvolutionMode())) {
+            LOGGER.warn("Table '{}' cannot be created because schema evolution is disabled.", tableName);
+            throw new SQLException("Cannot create table " + tableName + " because schema evolution is disabled");
+        }
+
+        Transaction transaction = session.beginTransaction();
+        try {
+            final String createSql = dialect.getCreateTableStatement(record, tableName);
+            LOGGER.trace("SQL: {}", createSql);
+            session.createNativeQuery(createSql, Object.class).executeUpdate();
+            transaction.commit();
+        }
+        catch (Exception e) {
+            transaction.rollback();
+            throw e;
+        }
+
+        return readTable(tableName);
+    }
+
+    private TableDescriptor alterTableIfNeeded(String tableName, SinkRecordDescriptor record) throws SQLException {
```

Comment:
I wonder if we need to access db at all in case of disabled schema evolution as it might be costly operation. Maybe we need to modes - unsafe where do we expect the table is basically managed by Debezium only so no metadata are checked after the initial table creation or metadata loading and safe that will do the checks.